### PR TITLE
[MERGE] crm: make TestLeadAssignPerf deterministic

### PR DESCRIPTION
**Pull request by Xavier-Do**
Purpose
========

The performances tests in TestLeadAssignPerf have a margin for queries
because of some randomness in query counts.

Those margins avoid random failures but will hide small increments,
making the query count fail randomly in future build. This margin also
makes the update of query counts difficult.

This commit tries to identify the source of this randomness and proposes
some fixes.

Sources of randomness
=====================
Comparing different executions, the query can differ on two point:
1. During the flush, while updating the lead
1.1 the written user_id can vary (~1 additional query)
1.2 the written convert_date can changes (~4 additional queries)
1.3 the written date_last_stage_update can change (~1 additional query)
2. during _handle_salesmen_assignment
2.1 write/_message_auto_subscribe can change? (~1 additional query)

the point 1.2 and 1.3 can be easily reproduced, adding sleep,
especially a 0.1 sleep in convert_opportunity

Writing different values for date will lead to multiple execute
when flushing the records:
the orm cannot group records with different values

Proposed fixes
==============
A. Use `cr.now` instead of `fields.datetime.now`
Using the transaction time will avoid randomness linked to the change of
second during the transaction. This will fix 1.2 and 1.3

B. Sorting members
The members comes from a o2m and the order is not deterministic.
During the test with a subset of lead (10 instead of 100),
the two last members  ('Martin Sales Manager' and 'Orteil Sales Own')
can come in different orders, leading to different user_id set on lead.
Instead of 4 different users, the lead where sometimes dispatch on
5 different users with this setup. (5 queries in flush() instead of 4)
This is fixed by adding a complete order (adding id) on crm.team.members
This will solve 1.1

C. The leads are also ordered by id if they have the same probability,
since it shouldn't hurt and make the search more deterministic.

2.1 was fixed by B or C (or is not fixed) not sure anymore.

Links
=====

Task-2796579

### CODE REVIEW

File: addons/crm/tests/test_performances.py
Comment (xmo-odoo) [inline addons/crm/tests/test_performances.py:95]: Is this supposed to remain here?

File: odoo/tools/speedscope.py
Comment (xmo-odoo) [inline odoo/tools/speedscope.py:95]: ???

File: addons/crm/tests/test_crm_lead_assignment.py
Comment (xmo-odoo) [inline addons/crm/tests/test_crm_lead_assignment.py:33]: Maybe we should just do that during tests in general? This log line exist specifically because "[deletions] leave no trace in database" (though arguably writes don't really either, a subsequent write to different fields of the same object will disappear them unless the information is specifically logged to the chatter).

For tests that just generates noise.

File: odoo/tools/speedscope.py
Comment (xmo-odoo) [inline odoo/tools/speedscope.py:175]: What does the trailing `/` mean?

**Review by xmo-odoo — APPROVED** [review]

**Comment by xmo-odoo** [comment]
Fix commit message: "determinist" (or "determinists") is not a word.

File: addons/crm/models/crm_team_member.py
Comment (ryv-odoo) [inline addons/crm/models/crm_team_member.py:124]: It will more clean to change the default order of the model (`crm.team.member`) instead, no?

File: addons/crm/tests/test_performances.py
Comment (Xavier-Do) [reply addons/crm/tests/test_performances.py:95]: No I need to remove that before merging :) I left it with the idea to run multibuilds on runbot and hopefully have a profile with all queries if a count breaks.

File: odoo/tools/speedscope.py
Comment (Xavier-Do) [reply odoo/tools/speedscope.py:95]: I will remove that but I would like to keep it one way or another. 
This output would be added if we have only sql, but in first position it is not a good idea, but it was convenient for me to spot the sources of indeterminism. I'll change that.

Here in my results, I can compare if two fonctions call have the same number of queries by looking at the execution time.

File: addons/crm/tests/test_crm_lead_assignment.py
Comment (Xavier-Do) [reply addons/crm/tests/test_crm_lead_assignment.py:33]: I agree this could maybe be done for all tests. But I don't know if many tests will unlink data. We could check the logs.

File: addons/crm/models/crm_team_member.py
Comment (Xavier-Do) [reply addons/crm/models/crm_team_member.py:124]: Yes you are right! ;)

File: addons/crm/tests/test_crm_lead_assignment.py
Comment (xmo-odoo) [reply addons/crm/tests/test_crm_lead_assignment.py:33]: > I agree this could maybe be done for all tests. But I don't know if many tests will unlink data. We could check the logs.

Looking at the latest master, at_install reports "more than 1000", and the post_install batches report 651, 518, 347, 161, 49, and 335 (after some more thinking I figure the entire loading process could do without this).

I guess it might be valuable when trying to debug tours but that seems somewhat unlikely, I don't know that people would notice an unlink having gone missing.

An other possibility might be for the runbot to just run with that specific logger disabled :D

Comment (Xavier-Do) [reply addons/crm/tests/test_crm_lead_assignment.py:33]: > An other possibility might be for the runbot to just run with that specific logger disabled :D

Sneaky
I prefer to keep the runbot behaviour as closes as possible to the local one. We could maybe make this log conditionnal in models.py.

Comment (xmo-odoo) [reply addons/crm/tests/test_crm_lead_assignment.py:33]: > > An other possibility might be for the runbot to just run with that specific logger disabled :D
> 
> Sneaky I prefer to keep the runbot behaviour as closes as possible to the local one. We could maybe make this log conditionnal in models.py.

I mean sure, but manipulating `--log-handler` is not exactly a biggie. Especially since the runbot already has some special handling of a bespoke log level and stuff.

File: addons/crm/tests/test_crm_lead_convert.py
Comment (tde-banana-odoo) [inline addons/crm/tests/test_crm_lead_convert.py:495]: Note that we could check date_open and date_conversion values prior to calling ``action_apply``, just to be sure it is updated by the method itself.

File: addons/crm/tests/test_performances.py
Comment (tde-banana-odoo) [inline addons/crm/tests/test_performances.py]: Linting: In commit "make TestLeadAssignPerf deterministic" you have some changes in test_performances.py that could be removed, as anyway you alter that file in the final commit "update query count".

File: addons/crm/models/crm_lead.py
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:324]: Side note: we could also update date_closed computation in crm_lead file, just to be coherent (not having any impact on tests but while we are cleaning it, maybe we could cleanup all datetime fields).

**Review by tde-banana-odoo — COMMENTED** [review]
Code LGTM :) some small comments while reading the diff.

Did not check the tooling code, as anyway you are more expert than me on that point :D .

Could you include task ID (2796579) now that we have one ?

Thanks !

**Comment by Xavier-Do** [comment]
Sorry for the noise, I was making some tests with codeowners

**Comment by tde-banana-odoo** [comment]
@Xavier-Do still some conflict to solve on performances ?

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Pull request closed without merging**
